org.osgi.test.junit4/src/test/java/org/osgi/test/junit4/context/BundleContextRuleTest.java

```diff
@@ -230,14 +232,37 @@ public void cleansUpGottenServiceObjects() throws Exception {
 
 	@Test
 	public void closeableBundleContext_handlesSelectedMethodsOfObject() throws Exception {
-		BundleContext upstream = FrameworkUtil.getBundle(getClass())
-			.getBundleContext();
+		Bundle upstream = FrameworkUtil.getBundle(getClass());
+		BundleContext upstreamContext = upstream.getBundleContext();
 		BundleContext closeableBC = CloseableBundleContext.proxy(upstream);
 
 		assertThat(closeableBC.toString()).as("toString")
 			.startsWith(CloseableBundleContext.class.getSimpleName())
-			.contains(upstream.toString());
+			.contains(upstreamContext.toString());
 		assertThat(closeableBC.hashCode()).as("hashcode")
-			.isEqualTo(upstream.hashCode());
+			.isEqualTo(upstreamContext.hashCode());
+	}
+
+	@Test
+	public void closeableBundleContext_survivesBundleRestart() throws BundleException {
+		Bundle upstream = FrameworkUtil.getBundle(getClass());
+		BundleContext upstreamContext = upstream.getBundleContext();
+		BundleContext closeableBC = CloseableBundleContext.proxy(upstream);
+
+		// Should not throw an exception
+		closeableBC.getServiceReference(Object.class);
+
+		upstream.stop();
+
+		try {
+			// Should throw an exception
+			closeableBC.getServiceReference(Object.class);
+			fail("Expected an IllegalStateException");
+		} catch (IllegalStateException ise) {}
+
+		upstream.start();
+
+		// Should not throw an exception
+		closeableBC.getServiceReference(Object.class);
```

Comment:
What is going to happen here when you call `close()`? It seems to me the following:

* First call to `getServiceReference()` is on the original `BundleContext` BC1. CBC will cache the service reference obtained thereby so that it can unget it when close() is called..
* Restart will invalidate the original `BundleContext` BC1 (and the service reference obtained above) and create a new one BC2.
* Second call to `getServiceReference()` will be on BC2. CBC will cache this service reference too.
*  When `close()` is called, it will try and unget both of the services references on BC2.

This seems problematic to me. I think that, if we're going to go down this route (which I'm still not 100% conviced of) the CloseableBundleContext needs to invalidate all of its cached references when the Bundle stops..